tests/test_vcompressor.py

```diff
@@ -313,6 +380,29 @@ def test_file_info(self):
 
             self.assertEqual(origin_json_data["file_info"], dup_json_data["file_info"])
 
+    def test_fee(self):
```

Comment:
I'm wondering why should we need to make sure the order of the events are kept the same. It seemed that even if the order is different, the `vizviewer` shows the same results.

As for the precision of the timestamp, in a single thread, the resolution is `20ns` as in the code
```c
static double get_ts(struct ThreadInfo* info)
{
    double curr_ts = get_system_ts();
    if (curr_ts <= info->prev_ts) {
        // We use artificial timestamp to avoid timestamp conflict.
        // 20 ns should be a safe granularity because that's normally
        // how long clock_gettime() takes.
        // It's possible to have three same timestamp in a row so we
        // need to check if curr_ts <= prev_ts instead of ==
        curr_ts = info->prev_ts + 20;
    }
    info->prev_ts = curr_ts;
    return curr_ts;
}
```
Based on _Sampling Theory_, a signal can be exactly reproduced if it is sampled at a frequency F, where F is greater than twice the maximum frequency in the signal(I'm not sure if the theory is suitable here). I think there won't be the same timestamps in a single thread.

Reply:
It's not about sampling theory. It's about the conversion between double and int, and the computing error of multiplication and division. The best way to have a basic idea is to test it.

For the order of events. I'm not saying in json file the events should be stored in the same order. We should check that if A happens before B before compression, A will still be before B after the compression. As we are messing up timestamps quite a lot by doing some computation, we should have some sanity check.